torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+    ):
+        """
+        Lookup a cache entry for the given arguments. If none exists, dispatch
+        and cache the result (if the result is cache-able).
+        """
+        try:
+            key = self._cache_key(func, args, kwargs)
+        except _BypassDispatchCache as e:
+            FakeTensorMode.cache_bypasses[e.reason] += 1
+            return self._dispatch_impl(func, types, args, kwargs)
+
+        entry = FakeTensorMode.cache.get(key, None)
+        if entry is not None:
+            FakeTensorMode.cache_hits += 1
+            output = self._output_from_cache_entry(entry, func, args)
+        else:
+            FakeTensorMode.cache_misses += 1
+            output = self._dispatch_impl(func, types, args, kwargs)
+            try:
+                self._create_cache_entry(key, args, output)
+            except _BypassDispatchCache as e:
+                FakeTensorMode.cache_bypasses[e.reason] += 1
+
+        return output
+
+    def _cache_key(
+        self, func: OpOverload, args: Tuple[Any], kwargs: Dict[str, Any]
+    ) -> _DispatchCacheKey:
+        """
+        Create a cache key given the dispatch args. Raise _BypassDispatchCache
+        for any situation that precludes caching.
+        """
+        # Avoid caching for any ops that would require a more sophisticated
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
+            raise _BypassDispatchCache("data dependent output")
+
+        if torch.Tag.dynamic_output_shape in func.tags:
+            raise _BypassDispatchCache("dynamic output shape")
+
+        if torch.Tag.inplace_view in func.tags:
+            raise _BypassDispatchCache("inplace view")
+
+        if func == aten._unsafe_view.default:
+            raise _BypassDispatchCache("unsafe view")
+
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
+
+        if func in self.lift_fns:
+            raise _BypassDispatchCache("lift")
+
+        # In order to handle storage aliasing, we need to establish the alias
+        # for any view op on a cache hit. But CompositeImplicitAutograd ops may
+        # or may not alias the input, so just punt on caching these.
+        if func.is_view and torch._C._dispatch_has_kernel_for_dispatch_key(
+            func.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        ):
+            raise _BypassDispatchCache("CompositeImplicitAutograd")
+
+        key_values = (
+            func,
+            # Translate any FakeTensor args to metadata.
+            self._translate_arg(args),
+            self._translate_arg(kwargs),
+            # Capture the default_dtype mode since that can affect the output tensor,
+            # e.g., when operating on constant float values.
+            torch.get_default_dtype(),
+            # The disallowance of dynamic shapes can affect caching, e.g., a change of
```

Comment:
There was one failing unit test that exposed the following special-case. The test was expecting an exception to be thrown after changing this mode, but I was getting a cache hit (correctly according to the dispatch args), and therefore didn't raise. This is another thing that makes me question whether I've really found every case where behavior is subject to some external mode or global.

Reply:
That is pretty strange. Link to test ?

Reply:
@eellison : `test_opcheck_opinfo_NumpyNMSCustomOp_cuda_float32`
Here I think: https://github.com/pytorch/pytorch/blob/main/test/test_custom_ops.py#L326-L330

Reply:
Perhaps there's a more direct top-level setting to capture whether dynamic shapes are allowed or disallowed?